pkg/migrator/core.go

```diff
@@ -182,57 +173,58 @@ func (m *migrator) migrateList(l runtime.Object) error {
 	return utilerrors.NewAggregate(errors)
 }
 
-func (m *migrator) worker(stop <-chan struct{}, workc <-chan runtime.Object, errc chan<- error) {
+func (m *migrator) worker(stop <-chan struct{}, workc <-chan *unstructured.Unstructured, errc chan<- error) {
+	var err error
 	for item := range workc {
+	Retry:
 		for {
-			err := m.try(item)
+			item, err = m.try(item)
 			if err == nil {
 				break
 			}
 			if canRetry(err) {
+				if seconds, delay := errors.SuggestsClientDelay(err); delay {
+					time.Sleep(time.Duration(seconds) * time.Second)
+				}
 				continue
 			}
 			select {
 			case errc <- err:
-				continue
+				break Retry
 			case <-stop:
 				return
 			}
 		}
 	}
 }
 
-// try tries to migrate the single object by GET and then PUT.
-func (m *migrator) try(work runtime.Object) error {
-	namespace, err := metadataAccessor.Namespace(work)
+// try tries to migrate the single object by PUT. If the PUT fails due to
+// conflicts, the function refreshes the object. If `try` returns an error, the
+// caller may retry with the returned object.
+func (m *migrator) try(item *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	namespace, err := metadataAccessor.Namespace(item)
 	if err != nil {
-		return err
+		return item, err
 	}
-	name, err := metadataAccessor.Name(work)
+	name, err := metadataAccessor.Name(item)
 	if err != nil {
-		return err
+		return item, err
 	}
-	get := m.get(namespace, name)
-	data, err := get.Raw()
-	if err != nil {
-		return retriable(get.Error())
+	_, err = m.put(namespace, name, item)
+	if err == nil {
+		return nil, nil
 	}
-
-	put := m.put(namespace, name, data)
-	if err := put.Error(); err != nil {
-		return retriable(put.Error())
+	if err != nil && !errors.IsConflict(err) {
+		// Return the original item in case the caller decided to retry.
+		return item, err
 	}
-
-	// TODO: The oc migrator tried to count if the update causes changes.
-	// To do so it needs to decode the PUT response and extract the
-	// resourceVersion.
+	return m.get(namespace, name)
```

Comment:
so if there is a conflict, this will return what's stored without retry, but it is possible what's stored there is still the old version

Reply:
This is a bug. I fixed the logic. I'll add unit test when the dynamic client is fixed.